contracts/Staking.sol

```diff
@@ -0,0 +1,45 @@
+pragma solidity 0.6.2;
+
+import "./vendor/SafeMath.sol";
+import "./StorageManager.sol";
+
+contract Staking {
+
+    using SafeMath for uint256;
+
+    StorageManager storageManager;
+
+    mapping(address => uint256) internal amountStaked;
+
+    event Staked(address indexed user, uint256 amount, uint256 total, bytes data);
+    event Unstaked(address indexed user, uint256 amount, uint256 total, bytes data);
+
+    constructor(address _storageManager) public {
+        storageManager = StorageManager(_storageManager);
+    }
+
+    function stake(bytes memory data) public payable {
+        stakeFor(msg.sender, data);
+    }
+
+    function stakeFor(address user, bytes memory data) public payable {
+        amountStaked[user] = amountStaked[user].add(msg.value);
+        emit Staked(msg.sender, msg.value, amountStaked[user], data);
+    }
+
+    function unstake(uint256 amount, bytes memory data) public {
+        // only allow unstake if there is no utilized capacity
+        require(storageManager.hasUtilizedCapacity(msg.sender));
+        amountStaked[msg.sender] = amountStaked[msg.sender].sub(amount);
+        emit Unstaked(msg.sender, amount, amountStaked[msg.sender], data);
+    }
+
+    function totalStaked() public view returns (uint256) {
+        return totalStakedFor(msg.sender);
```

Comment:
I think the `totalStaked` without parameter should return the TOTAL staked in the complete contract, according to the ERC900.  So we may beed an attribute to store this and add or subtract when corresponds. I think its good to have that indicator in the contract directly.

Reply:
Done